Add Isar output mode

edoput opened the pull request:
Checklist: 

- [ ] The wiki is updated in accordance with the changes in this PR. For example: syntax changes, semantics changes, VerCors flags changes, etc.

# PR description

This PR introduces a syntactic transformation from PVL to Isabelle. This allows for translating PVL's ADTs to a locale to prove they are consistent. The transformation is available as a command from vct, see example.

```
bin/vct --isar --isar-output=foo.thy res/universal/res/adt/*.pvl 
```

This creates a new file `foo.thy` pre-filled with a locale for each ADT. The user can then provide a witness for each locale to prove the corresponding ADT consistent.

Commit:
Add Isar mode to VerCors

this commit adds a print mode to output Isar syntax

Commit:
add Isabelle type syntax to types pretty printer

Commit:
add foldr to Doc

fold is defined as foldl which is left associative but I need the right
associative version for printing function types in ML

Commit:
implement pretty printing of ADT as locale

Commit:
fix locale declaration syntax error

Commit:
fix more issue with locale syntax

Commit:
implement basic binder syntax

Commit:
implement logical operators synta

Commit:
fix locale layout issues with fixes and asssumes

superaxander committed:
setup Isar output and required transformation passes

Commit:
only print the ADT when working in Isar mode

Commit:
[fix][Isar] only print the name of a variable and not its type

Commit:
correctly print functions that are defined outside of an ADT

Commit:
implement Isar printng for PVL functions and syntax

Commit:
implement isar output for rational division

Commit:
prelude: import rational numbers and add a type for references

Commit:
literal syntax fixes

Commit:
translate PVL sizes functions to be of type int in place of nat

Commit:
use nice <--> binder in place of = when comparing booleans

Commit:
typo

Commit:
fixes map constructor syntax and lookup translation

Commit:
uses HOL implication instead of meta-implication

Commit:
typo: oopsie

bobismijnnaam commented:
I converted your PR into a draft one. Feel free to un-draft it once you're planning on merging it soon.

superaxander commented on src/col/vct/col/ast/expr/literal/build/LiteralTupleImpl.scala:
```diff
@@ -15,6 +15,11 @@ trait LiteralTupleImpl[G] extends ExprImpl[G] with LiteralTupleOps[G] {
-    Group(Text("tuple<") <> Doc.args(ts) <> ">{" <> Doc.args(values) <> "}")
+  override def layout(implicit ctx: Ctx): Doc = {
+    ctx.syntax match {
+      case Ctx.Isar => Group(Text("(") <> Doc.args(ts) <> "," <> Doc.args(values) <>Text(")"))
```
I think this might not be the right syntax in Isabelle, `ts` is the sequence of types in the tuple

superaxander commented on src/col/vct/col/ast/expr/op/collection/SliceImpl.scala:
```diff
@@ -1,14 +1,18 @@
-    assoc(xs) <> "[" <> from <> ".." <> to <> "]"
+  override def layout(implicit ctx: Ctx): Doc = {
+    ctx.syntax match {
+      case Ctx.Isar => Text("nths") <+> assoc(xs) <+> "{" <> from <> ".." <> to <> "}"
```
I believe in Isabelle this is an inclusive range whereas in VerCors this is exclusive. 

In PVL
```
seq<int> s = [1,2,3];
s[1..2] == [2]
```

In Isabelle
```isabelle
nths [1::int,2,3] {1 .. 2} = [2, 3]
```

superaxander reviewed (changes requested):
I have two small comments

Commit:
[fix] conversion to Isabelle/HOL tuples, range not inclusive

Commit:
[isar] output of trigger should be optional

Commit:
implement conditional layout of triggers

Commit:
[fix] correctly typeset constants

Commit:
implement ColToIsar rewrite pass

Commit:
[isar][fix] PVL maps are finite, map them to finite maps and their
operations

Commit:
[isar][fix] PVL sets are finite, map them to finite sets and their
operations

Commit:
[fix] literal tuple syntax

Commit:
[isar][fix] correctly import finite multisets

also use the correct symbol for the cardinality

superaxander commented on src/col/vct/col/ast/lang/Isar/IsarDefinitionCommandImpl.scala:
```diff
@@ -0,0 +1,26 @@
+    Text("definition") <+> this.name <+> <::> <+> inner {
+      type_signature(this.args.map(_.t) :+ this.returnType)
+    } </> where <+/> inner {
+      // TODO why is _ not printed?
```
_ is not printed because ignored is unused here

edoput replied:
I was referring to the constant `ignore` that should be printed as `_` in the output. This is now obseleted as the variable ignored is dead and we print variable names as the name of the arguments to a definition.

superaxander commented on src/col/vct/col/ast/lang/Isar/IsarLocaleCommandImpl.scala:
```diff
@@ -0,0 +1,50 @@
+
+  def layout_assumes(implicit ctx: Ctx): Doc = {
+    Doc.stack(this.assumes.map { a =>
+      // TODO make a an ADT axiom
```
What’s the plan here?

edoput replied:
This refers to the plan of making a node for IsarAssumesCommand which we don't need as we can print pvl formulas in Isar style already.

superaxander commented on src/col/vct/col/ast/Node.scala:
```diff
@@ -59,6 +59,19 @@ import vct.col.ast.family.parregion._
 import vct.col.ast.family.pvlcommunicate._
 import vct.col.ast.family.seqrun._
 import vct.col.ast.family.signals._
+import vct.col.ast.lang.Isar.{
```
Replace with _ import please

superaxander commented on src/main/vct/main/modes/Isar.scala:
```diff
@@ -0,0 +1,40 @@
+    val result = Parsing.ofOptions(options, blameProvider)
+      .thenRun(Resolution.ofOptions(options, blameProvider))
+      .thenRun(Transformation.isarOfOptions(options)).thenRun(Output(
+        options.isarOutput.orElse(Some(Paths.get("a.java"))),
```
I guess this should be different default file name

superaxander commented on src/rewrite/vct/rewrite/ColToIsar.scala:
```diff
@@ -0,0 +1,383 @@
+            }
+          }
+
+          // NOTE type variables mapping is bound to typedef node
```
Pretty sure we dropped this type variable mapping thing so the notes can go away

superaxander commented on src/rewrite/vct/rewrite/ColToIsar.scala:
```diff
@@ -0,0 +1,383 @@
+
+        } else {
+          // typedef not required
+          // NOTE type variable mapping is bound to datatype node
```
Idem

superaxander reviewed (commented):
Looks good, please look at the comments I made and mark the PR as ready to review when you want it merged

Commit:
[fix] address review

Commit:
[fix] parenthesise adt function invocation when typesetting into Isar

Commit:
[fix] correctly typeset locale extensions

Commit:
[fix] compute both locale lexical scope and locale hierarchy

ADTs are translated to locales so we need to bind the lexical scope
of locales to the lexical scope of ADTs. Moreover in the interpretation
command we require the full list of parameters to the locales hierarchy.

superaxander commented on src/col/vct/col/ast/expr/literal/build/LiteralSetImpl.scala:
```diff
@@ -9,6 +9,14 @@ trait LiteralSetImpl[G] extends LiteralSetOps[G] {
+    ctx.syntax match {
+      case Ctx.Isar =>
+        if (values.isEmpty) { Group(Text("{}")) }
+        else { Group(Text("{") <> Doc.args(values) <> "}") }
```
Should be a finite set not a normal set

superaxander replied:
Also RangeSetImpl and RangeImpl (sets and lists respectively might need a translation)

superaxander commented on src/col/vct/col/ast/expr/op/collection/SizeImpl.scala:
```diff
@@ -9,5 +9,16 @@ trait SizeImpl[G] extends SizeOps[G] {
+          case TSeq(_) => Text("int(length") <+> obj <> ")"
+          case TSet(_) => Text("int(fcard") <+> obj <> ")"
+          case TBag(_) => Text("int(size") <+> obj <> ")"
+          case TMap(_, _) => Text("int(fcard (fmdom") <+> obj <> "))"
```
These should be `(int (length obj))` not `int (length obj)`

superaxander commented:
Also missing Isar pretty printing for `SeqSubscript`
